base/promotion.jl

```diff
@@ -102,37 +103,39 @@ end
 
 ## promotion mechanism ##
 
-promote_type()  = Bottom
-promote_type(T) = T
-promote_type(T, S, U, V...) = promote_type(T, promote_type(S, U, V...))
+promote_type()  = (@_pure_meta; Bottom)
+promote_type(T) = (@_pure_meta; T)
+promote_type(T, S, U, V...) = (@_pure_meta; promote_type(T, promote_type(S, U, V...)))
 
-promote_type(::Type{Bottom}, ::Type{Bottom}) = Bottom
-promote_type{T}(::Type{T}, ::Type{T}) = T
-promote_type{T}(::Type{T}, ::Type{Bottom}) = T
-promote_type{T}(::Type{Bottom}, ::Type{T}) = T
+promote_type(::Type{Bottom}, ::Type{Bottom}) = (@_pure_meta; Bottom)
+promote_type{T}(::Type{T}, ::Type{T}) = (@_pure_meta; T)
+promote_type{T}(::Type{T}, ::Type{Bottom}) = (@_pure_meta; T)
+promote_type{T}(::Type{Bottom}, ::Type{T}) = (@_pure_meta; T)
 
 # Try promote_rule in both orders. Typically only one is defined,
 # and there is a fallback returning Bottom below, so the common case is
 #   promote_type(T, S) =>
 #   promote_result(T, S, result, Bottom) =>
 #   typejoin(result, Bottom) => result
-promote_type{T,S}(::Type{T}, ::Type{S}) =
+function promote_type{T,S}(::Type{T}, ::Type{S})
+    @_pure_meta
```

Comment:
So what happens if I define a promote rule that ensures my custom type only ever has a concrete type parameter?  That would mean that this pure annotation is no longer correct, no?

```jl
@inline function Base.promote_rule{T,S<:Number}(::Type{Interval{T}}, ::Type{S})
    R = promote_type(T,S)
    isleaftype(R) || throw(ArgumentError("cannot promote $T and $S to a common leaf type"))
    Interval{R}
end
```

On second thought, this is probably better to enforce in the constructor.  But do we need to document that `promote_rule` needs to be a pure function now?  What happens if it's not?